lib/AST/ASTContext.cpp

```diff
@@ -4453,6 +4453,23 @@ ASTContext::getClangFunctionType(ArrayRef<AnyFunctionType::Param> params,
   return impl.Converter.getValue().getFunctionType(params, resultTy, trueRep);
 }
 
+void ASTContext::getSubstitutedClangTypesForTemplateFunction(
+    clang::FunctionTemplateDecl *func, SubstitutionMap subst,
+    SmallVectorImpl<clang::TemplateArgument> &convertedParams) {
+  auto &impl = getImpl();
+  if (!impl.Converter) {
+    auto *cml = getClangModuleLoader();
+    impl.Converter.emplace(*this, cml->getClangASTContext(), LangOpts.Target);
+  }
+
+  for (clang::NamedDecl *param : *func->getTemplateParameters()) {
+    auto templateParam = cast<clang::TemplateTypeParmDecl>(param);
+    auto replacement = subst.getReplacementTypes()[templateParam->getIndex()];
+    auto qualType = impl.Converter->convert(replacement);
+    convertedParams.push_back(clang::TemplateArgument(qualType));
+  }
```

Comment:
Please add some error handling or TODOs for the two failure modes where (1) we have a non-type template parameter and (2) the call to `convert` fails.

Arguably, if you aren't doing error handling for (1), there is no reason to take the `func` parameter at all, why not simply loop over `subst.getReplacementTypes()`? Is there a case where the result for `templateParam->getIndex()` will not be `0, 1, 2 ...`?

Reply:
I added a check in the importer for the first failure so, we should never receive TemplateParameterList with a non-type template parameter (I also added a comment about that in the loop). And I added a failure/error message for the second failure mode. 

You're right, passing replacement types in as an argument is better. Makes this method more generic/useful.